From 6411a5a8be62a37df0b59551381eaead2a3d820d Mon Sep 17 00:00:00 2001
From: Siddharth Manthan <siddharth_manthan@outlook.com>
Date: Fri, 7 May 2021 13:09:49 +0530
Subject: [PATCH 21/24] arm64: dts: qcom: msm8916-samsung-gt58: Add sound card

Additionally to pm8916 sound, gt58 uses single NXP TFA9895 audio amplifier.
Secondary mic is on gpio which has been forced on using pin control.

Signed-off-by: Siddharth Manthan <siddharth_manthan@outlook.com>
---
 .../boot/dts/qcom/msm8916-samsung-gt58.dts    | 47 +++++++++++++++++++
 1 file changed, 47 insertions(+)

diff --git a/arch/arm64/boot/dts/qcom/msm8916-samsung-gt58.dts b/arch/arm64/boot/dts/qcom/msm8916-samsung-gt58.dts
index 7b71da9e3e..1ed2a04a62 100644
--- a/arch/arm64/boot/dts/qcom/msm8916-samsung-gt58.dts
+++ b/arch/arm64/boot/dts/qcom/msm8916-samsung-gt58.dts
@@ -8,6 +8,26 @@ / {
 	model = "Samsung Galaxy Tab A 8.0 (2015)";
 	compatible = "samsung,gt58", "qcom,msm8916";
 
+	i2c-amplifier {
+		compatible = "i2c-gpio";
+		sda-gpios = <&msmgpio 55 (GPIO_ACTIVE_HIGH|GPIO_OPEN_DRAIN)>;
+		scl-gpios = <&msmgpio 56 (GPIO_ACTIVE_HIGH|GPIO_OPEN_DRAIN)>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&amp_i2c_default>;
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		speaker_codec: audio-codec@34 {
+			compatible = "nxp,tfa9895";
+			reg = <0x34>;
+			vddd-supply = <&pm8916_l5>;
+			sound-name-prefix = "Speaker";
+			#sound-dai-cells = <0>;
+		};
+	};
+
 	reg_5p4v: regulator-5p4v {
 		compatible = "regulator-fixed";
 		regulator-name = "vlcd_5p4v";
@@ -98,7 +118,25 @@ &mdss {
 	status = "okay";
 };
 
+&sound {
+
+	/* FIXME: Avoid powering secondary mic permanently */
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&cdc_pdm_lines_act &ext_sec_tlmm_lines_act &secondary_mic>;
+	pinctrl-1 = <&cdc_pdm_lines_sus &ext_sec_tlmm_lines_sus &secondary_mic>;
+
+	model = "samsung-a2015";
+};
+
 &msmgpio {
+	amp_i2c_default: amp-i2c-default {
+		pins = "gpio55", "gpio56";
+		function = "gpio";
+
+		drive-strength = <2>;
+		bias-disable;
+	};
+
 	buckbooster_en_default: buckbooster-en-default {
 		pins = "gpio8";
 		function = "gpio";
@@ -125,6 +163,15 @@ mdss_sleep: mdss-sleep {
 		};
 	};
 
+	secondary_mic: secondary-mic {
+		pins = "gpio98";
+		function = "gpio";
+
+		drive-strength = <2>;
+		bias-disable;
+		output-high;
+	};
+
 	reg_tsp_en_default: reg-tsp-en-default {
 		pins = "gpio73";
 		function = "gpio";
-- 
2.25.1

